strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/services/impl/ProxyRepositoryArtifactByteStreamsCopy.java

```diff
@@ -0,0 +1,280 @@
+import org.carlspring.strongbox.providers.io.RepositoryPath;
+import org.carlspring.strongbox.providers.layout.LayoutProvider;
+import org.carlspring.strongbox.providers.layout.LayoutProviderRegistry;
+import org.carlspring.strongbox.services.ArtifactByteStreamsCopyStrategy;
+import org.carlspring.strongbox.services.support.ArtifactByteStreamsCopyException;
+import org.carlspring.strongbox.storage.repository.Repository;
+import org.carlspring.strongbox.storage.repository.remote.RemoteRepository;
+import org.carlspring.strongbox.storage.repository.remote.heartbeat.RemoteRepositoryAlivenessCacheManager;
+
+import javax.inject.Inject;
+import javax.ws.rs.core.Response;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.net.URI;
+
+import org.apache.commons.lang.StringUtils;
+import org.apache.commons.lang3.BooleanUtils;
+import org.apache.commons.lang3.mutable.MutableObject;
+import org.apache.commons.lang3.time.StopWatch;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.stereotype.Component;
+
+/**
+ * @author Przemyslaw Fusik
+ */
+@Component
+public class ProxyRepositoryArtifactByteStreamsCopy
+        implements ArtifactByteStreamsCopyStrategy
+{
+
+    private static final Logger logger = LoggerFactory.getLogger(ProxyRepositoryArtifactByteStreamsCopy.class);
+
+    @Inject
+    private ConfigurationManager configurationManager;
+
+    @Inject
+    private RemoteRepositoryAlivenessCacheManager remoteRepositoryAlivenessCacheManager;
+
+    @Inject
+    private RestArtifactResolverFactory artifactResolverFactory;
+
+    @Inject
+    private LayoutProviderRegistry layoutProviderRegistry;
+
+    private ArtifactByteStreamsCopyStrategy simpleArtifactByteStreams = SimpleArtifactByteStreamsCopy.INSTANCE;
+
+    @Override
+    public long copy(final InputStream from,
+                     final OutputStream to,
+                     final RepositoryPath artifactPath)
+            throws IOException
+    {
+        final StopWatch stopWatch = new StopWatch();
+        stopWatch.start();
+
+        return copyWithOffset(from, to, artifactPath, stopWatch, 1, 0L, new MutableObject());
+    }
+
+    private long copyWithOffset(final InputStream from,
+                                final OutputStream to,
+                                final RepositoryPath artifactPath,
+                                final StopWatch stopWatch,
+                                final int attempts,
+                                final long currentOffset,
+                                final MutableObject<Boolean> rangeRequestSupported)
+            throws IOException
+    {
+        try
+        {
+            return currentOffset + copySimple(from, to, artifactPath);
+        }
+        catch (final ArtifactByteStreamsCopyException ex)
+        {
+            return retryCopyIfPossible(to, artifactPath, stopWatch, attempts, currentOffset + ex.getOffset(),
```

Comment:
Cosmetic: When there's a long list of arguments, could you please, always chop them to one per line?

Reply:
sure